File: static/js/queue.js
// queue.js — the import queue as the client sees it.
//
// One EventSource for the whole queue rather than one per job: browsers cap
// concurrent connections per origin at around six on HTTP/1.1 and the studio
// already spends most of that budget fetching stem WAVs, so twenty per-job
// streams would starve audio loading long before hitting any server limit.
//
// This module owns no DOM. It holds the latest snapshot and tells subscribers
// when it changes; catalog.js decides what that looks like.

const POLL_MS = 2000;
const MAX_SSE_ATTEMPTS = 6;

let snapshot = { running: null, queued: [], max_pending: 0, capacity_left: 0 };
let source = null;
let pollTimerId = null;
let attempt = 0;
const subscribers = new Set();

// ─── pure helpers (no DOM, no network -- the testable part) ───────────────────

export function queueCount(snap = snapshot) {
  return (snap.running ? 1 : 0) + (snap.queued?.length ?? 0);
}

export function ordinal(n) {
  const rem100 = n % 100;
  if (rem100 >= 11 && rem100 <= 13) return `${n}th`;
  const suffix = { 1: "st", 2: "nd", 3: "rd" }[n % 10] ?? "th";
  return `${n}${suffix}`;
}

/** Stage text for the running row. The backend stage already carries a
 *  percentage during separation ("Separating 42%"), so only append one when it
 *  does not, rather than rendering "Separating 42% 42%". */
export function runningLabel(job) {
  const stage = (job?.stage || "Working...").replace(/\.\.\.$/, "");
  if (/\d\s*%/.test(stage)) return stage;
  const pct = Math.round((job?.progress || 0) * 100);
  return pct > 0 ? `${stage} ${pct}%` : stage;
}

/** Per-job view state, keyed by job id, for whoever is drawing rows.
 *  Position counts the running job, so the first waiting job is 2nd in line. */
export function queueRowStates(snap = snapshot) {
  const rows = new Map();
  if (snap.running) {
    rows.set(snap.running.job_id, {
      state: "running",
      label: runningLabel(snap.running),
      progress: snap.running.progress || 0,
      position: 1,
    });
  }
  const offset = snap.running ? 2 : 1;
  (snap.queued ?? []).forEach((job, i) => {
    const place = i + offset;
    rows.set(job.job_id, {
      state: "waiting",
      label: `Queued - ${ordinal(place)} in line`,
      progress: 0,
      position: place,
    });
  });
  return rows;
}

// ─── state ───────────────────────────────────────────────────────────────────

export function getQueueSnapshot() {
  return snapshot;
}

export function onQueueChange(fn) {
  subscribers.add(fn);
  return () => subscribers.delete(fn);
}

function publish(next) {
  snapshot = next;
  for (const fn of subscribers) {
    try { fn(snapshot); } catch (e) { console.warn("[queue] subscriber failed:", e); }
  }
}

export async function refreshQueue() {
  try {
    const r = await fetch("/api/queue");
    if (!r.ok) return;
    publish(await r.json());
  } catch (e) {
    console.warn("[queue] refresh failed:", e);
  }
}

export async function cancelQueuedJob(jobId) {
  try {
    await fetch(`/api/jobs/${jobId}/cancel`, { method: "POST" });
  } catch (e) {
    console.warn("[queue] cancel failed:", e);
  }
  // Do not wait for the next frame: the row should go the moment it is clicked.
  await refreshQueue();
}

// ─── transport ───────────────────────────────────────────────────────────────

function startPolling() {
  if (pollTimerId) return;
  pollTimerId = setInterval(refreshQueue, POLL_MS);
  refreshQueue();
}

function stopPolling() {
  if (pollTimerId) {
    clearInterval(pollTimerId);
    pollTimerId = null;
  }
}

export function startQueueStream() {
  if (source) return;
  refreshQueue(); // first paint should not wait for the stream to connect

  const open = () => {
    const es = new EventSource("/api/queue/events");
    source = es;

    es.onmessage = (ev) => {
      attempt = 0;
      stopPolling(); // the stream is healthy again
      try {
        publish(JSON.parse(ev.data));
      } catch (e) {
        console.warn("[queue] bad frame:", e);
      }
    };

    es.onerror = () => {
      es.close();
      source = null;
      attempt += 1;
      if (attempt > MAX_SSE_ATTEMPTS) {
        // Give up on SSE and keep the view alive by polling, same fallback
        // shape as the per-job stream in job.js.
        startPolling();
        return;
      }
      setTimeout(open, 500 * Math.pow(2, attempt - 1)); // 0.5s .. 16s
    };
  };

  open();
}

export function stopQueueStream() {
  if (source) {
    source.close();
    source = null;
  }
  stopPolling();
}

File: app/api/queue.py
"""Aggregate view of the import queue.

One stream for every waiting job rather than one per job. A Map of per-job
EventSources would be simpler on the client, but browsers cap concurrent
connections per origin at around six on HTTP/1.1, and the studio already spends
most of that budget fetching stem WAVs -- twenty queue streams would starve
audio loading long before hitting the server-side connection cap.

The payload is deliberately compact (to_queue_state, not to_state). The
foreground import keeps its own /api/jobs/{id}/events stream, which is what
carries the full completion state the studio needs.
"""

from __future__ import annotations

import asyncio
import json
from collections.abc import AsyncIterator
from typing import Any

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

from app.api.events import _MAX_SSE_SECONDS, claim_sse_slot, release_sse_slot
from app.core.config import JOB_ID_RE, JOBS_DIR, MAX_PENDING_UPLOAD_JOBS, MAX_PENDING_URL_JOBS
from app.core.registry import get as registry_get
from app.core.registry import pending_count as registry_pending_count
from app.core.registry import persist as registry_persist
from app.pipeline import jobqueue

router = APIRouter(tags=["queue"])


def _snapshot() -> dict[str, Any]:
    """Current queue as the client sees it. Position is derived here rather than
    stored on the Job: it changes for every waiting job whenever the head is
    dequeued, so writing it through _set() would bump N versions and wake N
    streams for a value that is just a list index."""
    running_id, waiting = jobqueue.snapshot()

    running = None
    if running_id is not None:
        job = registry_get(running_id)
        if job is not None:
            running = job.to_queue_state()

    queued = []
    for position, job_id in enumerate(waiting):
        job = registry_get(job_id)
        if job is None:
            continue
        rec = job.to_queue_state()
        rec["position"] = position
        queued.append(rec)

    # Capacity is per kind: a waiting upload holds its source file on disk, a
    # waiting link holds nothing, so they are bounded separately.
    uploads_pending = registry_pending_count(uploads=True)
    urls_pending = registry_pending_count(uploads=False)
    return {
        "running": running,
        "queued": queued,
        # True when jobs were restored from a previous session and are waiting
        # for the user to start them.
        "paused": jobqueue.is_paused(),
        "max_pending_uploads": MAX_PENDING_UPLOAD_JOBS,
        "max_pending_urls": MAX_PENDING_URL_JOBS,
        "capacity_left_uploads": max(0, MAX_PENDING_UPLOAD_JOBS - uploads_pending),
        "capacity_left_urls": max(0, MAX_PENDING_URL_JOBS - urls_pending),
    }


def _fingerprint() -> tuple:
    """Cheap change detector: the ids in order plus each job's version counter,
    which _set() already bumps on every field write."""
    running_id, waiting = jobqueue.snapshot()
    ids = ([running_id] if running_id else []) + waiting
    out: list[object] = [jobqueue.is_paused()]
    for job_id in ids:
        job = registry_get(job_id)
        out.append((job_id, job.version if job is not None else -1))
    return tuple(out)


@router.post("/start")
def start_queue() -> dict[str, Any]:
    """Begin working through a queue restored from a previous session."""
    jobqueue.resume()
    return _snapshot()


class ReorderRequest(BaseModel):
    job_id: str
    # The job it should sit directly after. None means "move to the front",
    # which is also what a Move to top control sends.
    after: str | None = None


@router.post("/reorder")
def reorder_queue(payload: ReorderRequest) -> dict[str, Any]:
    """Move a waiting job. Returns the resulting queue, so a client whose drag
    raced a job finishing re-syncs from the answer instead of guessing."""
    if not JOB_ID_RE.match(payload.job_id):
        raise HTTPException(status_code=404, detail="job not found")
    if payload.after is not None and not JOB_ID_RE.match(payload.after):
        raise HTTPException(status_code=404, detail="job not found")
    if payload.job_id == payload.after:
        raise HTTPException(status_code=422, detail="a job cannot follow itself")
    if not jobqueue.reorder(payload.job_id, payload.after):
        # Already started or finished. Not an error the user can act on -- the
        # snapshot below tells the client what is true now.
        raise HTTPException(status_code=409, detail="that job is no longer waiting")
    registry_persist(JOBS_DIR)
    return _snapshot()


@router.get("")
def get_queue() -> dict[str, Any]:
    """The queue right now. Used for first paint and as the polling fallback
    when the stream cannot connect."""
    return _snapshot()


@router.get("/events")
async def queue_events() -> StreamingResponse:
    """SSE stream of the whole queue.

    Unlike the per-job stream this never self-closes on a terminal status: it
    outlives any individual job and is expected to stay open for the session,
    so only the 4 h ceiling ends it.
    """
    claim_sse_slot()

    async def stream() -> AsyncIterator[str]:
        try:
            last_fp: tuple | None = None
            keepalive_at = 0
            loop = asyncio.get_running_loop()
            deadline = loop.time() + _MAX_SSE_SECONDS
            while loop.time() < deadline:
                fp = _fingerprint()
                if fp != last_fp:
                    snapshot = _snapshot()
                    if _fingerprint() != fp:
                        # _set() landed mid-serialize; the snapshot could mix
                        # pre- and post-write fields. Re-read next tick rather
                        # than emit a torn frame (same guard as job_events).
                        continue
                    yield f"data: {json.dumps(snapshot)}\n\n"
                    last_fp = fp
                    keepalive_at = 0
                keepalive_at += 1
                if keepalive_at >= 60:  # ~15 s at the poll interval below
                    yield ": keepalive\n\n"
                    keepalive_at = 0
                await asyncio.sleep(0.25)
        finally:
            release_sse_slot()

    return StreamingResponse(
        stream(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )
